pennylane/ops/qubit.py

```diff
@@ -565,6 +567,96 @@ def decomposition(phi, theta, omega, wires):
+
+    .. math::
+
+        RP(\theta, P) = \exp(-i \frac{\theta}{2} P)
+
+    **Details:**
+
+    * Number of wires: Any
+    * Number of parameters: 2 (1 differentiable parameter)
+    * Gradient recipe: :math:`\frac{d}{d\theta}f(RP(\theta)) = \frac{1}{2}\left[f(RP(\theta +\pi/2)) - f(RP(\theta-\pi/2))\right]`
+      where :math:`f` is an expectation value depending on :math:`RP(\theta)`.
+
+    .. note::
+
+        If the ``PauliRot`` gate is not supported on the targeted device, PennyLane
+        will attempt to decompose the gate using :class:`~.RX`, :class:`~.Hadamard`, :class:`~.RZ`
+        and :class:`~.CNOT` gates.
+
+    Args:
+        theta (float): rotation angle :math:`\theta`
+        pauli_word (string): the Pauli word defining the rotation
+        wires (Sequence[int] or int): the wire the operation acts on
+    """
+    num_params = 2
+    num_wires = Any
+    do_check_domain = False
+    par_domain = "R"
+    grad_method = "A"
+
+    _PAULI_CONJUGATION_MATRICES = {
+        "X": Hadamard._matrix(),
+        "Y": RX._matrix(np.pi / 2),
+        "Z": np.eye(2),
+    }
+
+    @staticmethod
+    def _matrix(*params):
+        theta = params[0]
+        pauli_word = params[1]
+        # We keep imaginary wires to be able to reconstruct the position of the identities
+        active_wires, active_gates = zip(
+            *[(wire, gate) for wire, gate in enumerate(pauli_word) if gate != "I"]
+        )
+
+        # we first create the multi-Z rotation
+        multi_Z_rot_eigs = np.exp(-1j * theta / 2 * pauli_eigs(len(active_gates)))
+        multi_Z_rot_matrix = np.diag(multi_Z_rot_eigs)
+
+        # now we conjugate with Hadamard and RX to create the Pauli string
+        conjugation_matrix = functools.reduce(
+            np.kron, [PauliRot._PAULI_CONJUGATION_MATRICES[gate] for gate in active_gates],
+        )
+
+        return expand_matrix(
+            conjugation_matrix.T.conj() @ multi_Z_rot_matrix @ conjugation_matrix,
+            active_wires,
+            list(range(len(pauli_word))),
+        )
+
+    @staticmethod
+    @template
+    def decomposition(theta, pauli_word, wires):
+        active_wires, active_gates = zip(
+            *[(wire, gate) for wire, gate in zip(wires, pauli_word) if gate != "I"]
+        )
+
+        for wire, gate in zip(active_wires, active_gates):
+            if gate == "X":
+                Hadamard(wires=[wire])
+            elif gate == "Y":
+                RX(np.pi / 2, wires=[wire])
+
+        for i in range(len(active_wires) - 1, 0, -1):
+            CNOT(wires=[active_wires[i], active_wires[i - 1]])
+
+        RZ(theta, wires=active_wires[0])
+
+        for i in range(len(active_wires) - 1):
+            CNOT(wires=[active_wires[i + 1], active_wires[i]])
+
+        for wire, gate in zip(active_wires, active_gates):
+            if gate == "X":
+                Hadamard(wires=[wire])
+            elif gate == "Y":
+                RX(-np.pi / 2, wires=[wire])
```

Comment:
I imagine, if timed, that the decomposition approach to applying the rotation will be much faster than the kronecker product/expand matrix method. In fact, the latter probably scales really badly with increasing number of qubits on the device?

Reply:
I will do some benchmarks. Technically, when you do the CNOTs you're building the eigenvalue matrix that I already know a priori (and directly construct in the code), so I expected this to be faster than a lot of CNOTs.